go/enclave/components/batch_executor.go

```diff
@@ -32,23 +35,69 @@ type batchExecutor struct {
 	crossChainProcessors *crosschain.Processors
 	genesis              *genesis.Genesis
 	logger               gethlog.Logger
+	gasOracle            gas.Oracle
 	chainConfig          *params.ChainConfig
 
 	// stateDBMutex - used to protect calls to stateDB.Commit as it is not safe for async access.
 	stateDBMutex sync.Mutex
 }
 
-func NewBatchExecutor(storage storage.Storage, cc *crosschain.Processors, genesis *genesis.Genesis, chainConfig *params.ChainConfig, logger gethlog.Logger) BatchExecutor {
+func NewBatchExecutor(
+	storage storage.Storage,
+	cc *crosschain.Processors,
+	genesis *genesis.Genesis,
+	gasOracle gas.Oracle,
+	chainConfig *params.ChainConfig,
+	logger gethlog.Logger,
+) BatchExecutor {
 	return &batchExecutor{
 		storage:              storage,
 		crossChainProcessors: cc,
 		genesis:              genesis,
 		chainConfig:          chainConfig,
 		logger:               logger,
+		gasOracle:            gasOracle,
 		stateDBMutex:         sync.Mutex{},
 	}
 }
 
+// payL1Fees - this function modifies the state db according to the transactions contained within the batch context
+// in order to substract gas fees from the balance. It returns a list of the transactions that have prepaid for their L1
+// publishing costs.
+func (executor *batchExecutor) payL1Fees(stateDB *state.StateDB, context *BatchExecutionContext) common.L2Transactions {
+	transactions := make(common.L2Transactions, 0)
+	block, _ := executor.storage.FetchBlock(context.BlockPtr)
+
+	for _, tx := range context.Transactions {
+		sender, err := core.GetAuthenticatedSender(context.ChainConfig.ChainID.Int64(), tx)
+		if err != nil {
+			executor.logger.Warn("Unable to extract sender for tx", log.TxKey, tx.Hash())
+			continue
+		}
+		accBalance := stateDB.GetBalance(*sender)
+
+		cost, err := executor.gasOracle.GetGasCostForTx(tx, block)
+		if err != nil {
+			executor.logger.Warn("Unable to get gas cost for tx", log.TxKey, tx.Hash(), log.ErrKey, err)
+			continue
+		}
+
+		isFreeTransaction := tx.GasFeeCap().Cmp(gethcommon.Big0) == 0
+		isFreeTransaction = isFreeTransaction && tx.GasPrice().Cmp(gethcommon.Big0) == 0
+
+		if !isFreeTransaction {
+			if accBalance.Cmp(cost) == -1 {
+				executor.logger.Info("insufficient account balance for tx", log.TxKey, tx.Hash(), "addr", sender.Hex())
+				continue
+			}
+			stateDB.SubBalance(*sender, cost)
+			stateDB.AddBalance(context.Creator, cost)
```

Comment:
todo - add refund for transactions that can't be included in a batch because the gas limit was hit